### PR TITLE
Virtual Hub : Create module

### PR DESCRIPTION
**Pull request by MrMCake**
# Change

- Added module Virtual Hub
- Added child modules route table & connection

> Note: The module needs a virtual WAN deployed. As we don't have one constantly - this module will not work in CARML until we implemented a new approach for dependencies (similar to the Gateway Connection module)

Pipeline reference
[Network: Virtual Hubs](https://github.com/MrMCake/ResourceModules/actions/workflows/ms.network.virtualHubs.yml)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (Wiki)

## Checklist

- [ ] I'm sure there are no other open Pull Requests for the same update/change
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (readme)
- [ ] I did format my code

### CODE REVIEW

File: arm/Microsoft.Network/virtualHubs/.parameters/min.parameters.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "name": {
+            "value": "sxx-az-vHub-min-001"
```
Comment (eriqua) [inline arm/Microsoft.Network/virtualHubs/.parameters/min.parameters.json:6]:
```suggestion
            "value": "sxx-az-vhub-min-001"
```

File: .azuredevops/modulePipelines/ms.network.virtualHubs.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
+name: 'Network: Virtual Hubs'
```
Comment (MrMCake) [inline .azuredevops/modulePipelines/ms.network.virtualHubs.yml:1]: Add missing version file

**Review by rahalan — DISMISSED** [review]

File: .azuredevops/modulePipelines/ms.network.virtualHubs.yml (hunk 2 above)
Comment (MrMCake) [reply .azuredevops/modulePipelines/ms.network.virtualHubs.yml:1]: done

File: arm/Microsoft.Network/virtualHubs/.parameters/min.parameters.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "name": {
+            "value": "sxx-az-vhub-min-001"
+        },
+        "addressPrefix": {
+            "value": "10.0.0.0/16"
+        },
+        "virtualWanId": {
+            "value": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/virtualWans/sxx-az-vw-x-001"
```
Comment (MariusStorhaug) [inline arm/Microsoft.Network/virtualHubs/.parameters/min.parameters.json:12]: should be an adp resource, or even using the tokens from the other PR

File: arm/Microsoft.Network/virtualHubs/.parameters/parameters.json [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "name": {
+            "value": "sxx-az-vhub-x-001"
+        },
+        "addressPrefix": {
+            "value": "10.1.0.0/16"
+        },
+        "virtualWanId": {
+            "value": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/virtualWans/sxx-az-vw-x-001"
```
Comment (MariusStorhaug) [inline arm/Microsoft.Network/virtualHubs/.parameters/parameters.json:12]: should be an adp resource, or even using the tokens from the other PR

File: arm/Microsoft.Network/virtualHubs/.parameters/parameters.json [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+            ]
+        },
+        "hubVirtualNetworkConnections": {
+            "value": [
+                {
+                    "name": "connection1",
+                    "remoteVirtualNetworkId": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/virtualNetworks/adp-sxx-az-vnet-x-vhub",
+                    "routingConfiguration": {
+                        "associatedRouteTable": {
+                            "id": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/virtualHubs/sxx-az-vHub-x-001/hubRouteTables/routeTable1"
```
Comment (MariusStorhaug) [inline arm/Microsoft.Network/virtualHubs/.parameters/parameters.json:28]: should be an adp resource, or even using the tokens from the other PR

File: arm/Microsoft.Network/virtualHubs/.parameters/parameters.json [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+                        },
+                        "propagatedRouteTables": {
+                            "ids": [
+                                {
+                                    "id": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/virtualHubs/sxx-az-vHub-x-001/hubRouteTables/routeTable1"
```
Comment (MariusStorhaug) [inline arm/Microsoft.Network/virtualHubs/.parameters/parameters.json:33]: should be an adp resource, or even using the tokens from the other PR

File: arm/Microsoft.Network/virtualHubs/deploy.bicep [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+param routeTableRoutes array = []
+
+@description('Optional. ID of the Security Partner Provider to link to')
+param securityPartnerProviderId string = ''
+
+@description('Optional. The Security Provider name.')
+param securityProviderName string = ''
+
+@description('Optional. The sku of this VirtualHub.')
+param sku string = 'Standard'
```
Comment (MariusStorhaug) [inline arm/Microsoft.Network/virtualHubs/deploy.bicep:44]: This input supports basic and standard. Cannot find it in any other doc than [azcli](https://docs.azure.cn/zh-cn/cli/ext/virtual-wan/network/vhub?view=azure-cli-latest#ext-virtual-wan-az-network-vhub-create) or [terraform azurerm registry](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/virtual_hub)



```suggestion
@allowed([
  'Basic'
  'Standard'
])
@description('Optional. The sku of this VirtualHub.')
param sku string = 'Standard'
```

File: arm/Microsoft.Network/virtualHubs/readme.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+| `hubRouteTables` | _[hubRouteTables](hubRouteTables/readme.md)_ array | `[]` |  | Optional. Route tables to create for the virtual hub. |
+| `hubVirtualNetworkConnections` | _[hubVirtualNetworkConnections](hubVirtualNetworkConnections/readme.md)_ array | `[]` |  | Optional. Virtual network connections to create for the virtual hub. |
+| `location` | string | `[resourceGroup().location]` |  | Optional. Location for all resources. |
+| `name` | string |  |  | Required. The virtual hub name. |
+| `p2SVpnGatewayId` | string |  |  | Optional. Resource ID of the Point-to-Site VPN Gateway to link to |
+| `preferredRoutingGateway` | string |  | `[ExpressRoute, None, VpnGateway, ]` | Optional. The preferred routing gateway types |
+| `routeTableRoutes` | array | `[]` |  | Optional. VirtualHub route tables |
+| `securityPartnerProviderId` | string |  |  | Optional. ID of the Security Partner Provider to link to |
+| `securityProviderName` | string |  |  | Optional. The Security Provider name. |
+| `sku` | string | `Standard` |  | Optional. The sku of this VirtualHub. |
```
Comment (MariusStorhaug) [inline arm/Microsoft.Network/virtualHubs/readme.md:31]:
```suggestion
| `sku` | string | `Standard` | `Basic`,`Standard` | Optional. The sku of this VirtualHub. |
```

**Review by MariusStorhaug — CHANGES_REQUESTED** [review]
No diagsettings,lock or RBAC?

File: arm/Microsoft.Network/virtualHubs/.parameters/min.parameters.json (hunk 3 above)
Comment (MrMCake) [reply arm/Microsoft.Network/virtualHubs/.parameters/min.parameters.json:12]: Nope. This module is the same as the VNET Connection. It won't be deployable in our environment

File: arm/Microsoft.Network/virtualHubs/.parameters/parameters.json (hunk 4 above)
Comment (MrMCake) [reply arm/Microsoft.Network/virtualHubs/.parameters/parameters.json:12]: Nope. This module is the same as the VNET Connection. It won't be deployable in our environment

File: arm/Microsoft.Network/virtualHubs/.parameters/parameters.json (hunk 5 above)
Comment (MrMCake) [reply arm/Microsoft.Network/virtualHubs/.parameters/parameters.json:28]: Nope. This module is the same as the VNET Connection. It won't be deployable in our environment

File: arm/Microsoft.Network/virtualHubs/.parameters/parameters.json (hunk 6 above)
Comment (MrMCake) [reply arm/Microsoft.Network/virtualHubs/.parameters/parameters.json:33]: Nope. This module is the same as the VNET Connection. It won't be deployable in our environment

**Review by MariusStorhaug — APPROVED** [review]
✅

**Pull request merged**